hy/compiler.py

```diff
@@ -791,55 +786,57 @@ def compile_try_expression(self, expr):
+                                                              name)
+            handlers.append(handler_results.stmts.pop())
         orelse = []
+        if expr and expr[0][0] == HySymbol("else"):
+            orelse = self._compile_branch(expr.pop(0)[1:])
+            orelse += ast.Assign(targets=[name],
+                                 value=orelse.force_expr,
+                                 lineno=expr.start_line,
+                                 col_offset=expr.start_column)
+            orelse += orelse.expr_as_stmt()
+            orelse = orelse.stmts
         finalbody = []
-        handlers = []
-        handler_results = Result()
-
-        for e in expr:
-            if not len(e):
-                raise HyTypeError(e, "Empty list not allowed in `try'")
-
-            if e[0] == HySymbol("except"):
-                handler_results += self._compile_catch_expression(e, name)
-                handlers.append(handler_results.stmts.pop())
-            elif e[0] == HySymbol("else"):
-                orelse = self.try_except_helper(e, HySymbol("else"), orelse)
-            elif e[0] == HySymbol("finally"):
-                finalbody = self.try_except_helper(e, HySymbol("finally"),
-                                                   finalbody)
-            else:
-                raise HyTypeError(e, "Unknown expression in `try'")
+        if expr and expr[0][0] == HySymbol("finally"):
+            finalbody = self._compile_branch(expr.pop(0)[1:])
+            finalbody += finalbody.expr_as_stmt()
+            finalbody = finalbody.stmts
+        if expr:
+            if expr[0][0] in ("except", "else", "finally"):
+                raise HyTypeError(expr, "Incorrect order "
+                                  "of `except'/`else'/`finally' in `try'")
+            raise HyTypeError(expr, "Unknown expression in `try'")
 
         # Using (else) without (except) is verboten!
         if orelse and not handlers:
             raise HyTypeError(
-                e,
+                expr,
                 "`try' cannot have `else' without `except'")
 
-        # (try) or (try BODY)
-        # Generate a default handler for Python >= 3.3 and pypy
-        if not handlers and not finalbody and not orelse:
-            handlers = [ast.ExceptHandler(
-                lineno=expr.start_line,
-                col_offset=expr.start_column,
-                type=None,
-                name=None,
-                body=[ast.Raise(lineno=expr.start_line,
-                                col_offset=expr.start_column)])]
-
         ret = handler_results
 
+        body += body.expr_as_stmt() if orelse else ast.Assign(
+            targets=[name],
+            value=body.force_expr,
+            lineno=expr.start_line,
+            col_offset=expr.start_column)
+
+        if not (handlers or finalbody):
+            # (try) or (try BODY)
+            # As is, this would be illegal under Python >= 3.3.
+            # So forget the `try` and just use the BODY.
```

Comment:
My own thinking is that this behavior, albeit possibly surprising, will be useful for situations such as the programmer commenting out all the `except` clauses, or accumulating `except` clauses programatically. In general, it seems good for Hy to provide base cases of this kind that Python doesn't. I'm not really sure, though.